### PR TITLE
gh-102856: Initial implementation of PEP 701

### PR DESCRIPTION
**Pull request by pablogsal**
* Issue: gh-102856

### CODE REVIEW

File: Parser/tokenizer.h [hunk 1]
```diff
@@ -33,6 +33,35 @@ struct token {
...
+    int bracket_mark_index;
+
+    char f_string_quote;
+    int f_string_quote_size;
+    int f_string_raw;
+    const char* f_string_start;
+    const char* f_string_multi_line_start;
+
+    int last_expr_size;
+    int last_expr_end;
```
Comment (Eclips4) [inline Parser/tokenizer.h]:
```suggestion
    Py_ssize_t last_expr_size;
    Py_ssize_t last_expr_end;
```
CI/CD says about conversion size_t to int (and possible  loss of data)
I think, we can safely change types of this fields from `int` to `Py_ssize_t`.
If so, the warnings will go away.

File: Parser/tokenizer.h [hunk 2]
```diff
@@ -57,6 +86,8 @@ struct tok_state {
     int lineno;         /* Current line number */
     int first_lineno;   /* First line of a single line or multi line string
                            expression (cf. issue 16806) */
+    int fstring_first_constant_lineno; /* First line number of a single line or multiline
```
Comment (pablogsal) [inline Parser/tokenizer.h]: @lysnikolaou @isidentical shouldn't this be in the `_tokenizer_mode` struct?

File: Grammar/python.gram [hunk 3]
```diff
@@ -877,7 +877,29 @@ lambda_param[arg_ty]: a=NAME { _PyAST_arg(a->v.Name.id, NULL, NULL, EXTRA) }
 # LITERALS
 # ========
 
-strings[expr_ty] (memo): a=STRING+ { _PyPegen_concatenate_strings(p, a) }
+fstring_middle[expr_ty]:
+    | fstring_replacement_field
+    | t=FSTRING_MIDDLE { _PyPegen_constant_from_token(p, t) }
+# There are some shenanigans with the fstring_format_spec: Don't try to put it in its own rule
```
Comment (pablogsal) [inline Grammar/python.gram]: @lysnikolaou @isidentical Do you know if this is still true?

Comment (lysnikolaou) [reply Grammar/python.gram]: Checking now.

Comment (lysnikolaou) [reply Grammar/python.gram]: Turns out it isn't. Pushed a commit that does a re-organizing of these rules to your branch directly.

Comment (pablogsal) [reply Grammar/python.gram]: Fantastic!

**Comment by sunmy2019** [comment]
One issue is that, with current grammar
```py
f"{lambda x:{123}}"
```
will be recognized as a valid lambda, but
```py
f"{lambda x: {123}}"
f"{lambda x:{123} }"
```
won't. It definitely confuses the users.

I can't figure out an elegant way to fix this under current tokens. Since the info of `in_format_spec` only exists when the token is being tokenized, then the information is lost when exiting that 

One workaround is to emit an empty fstring_middle to prevent any further match by the `lambdef`.

Another workaround is to add 2 tokens: `FSTRING_REPLACEMENT_FIELD_START/END`, this preserves the `in_format_spec` info when passed to the parser.

**Comment by pablogsal** [comment]
@sunmy2019 the changes may also make `test_tokenize.TestRoundtrip.test_random_files` fail for some cases, but that may be an older failure.

**Comment by sunmy2019** [comment]
> @sunmy2019 the changes may also make `test_tokenize.TestRoundtrip.test_random_files` fail for some cases, but that may be an older failure.

I ran cpu heavy tests yeterday, and found this failure. See here: https://github.com/pablogsal/cpython/pull/67#issuecomment-1505840208

Both the tokenize and the untokenize function needs a rewrite.

**Comment by isidentical** [comment]
> I cannot reproduce in my mac machine. Maybe someone has more luck with a Linux system

No luck on my side either (with a Linux machine + debug build + refleaks) for `test_ast`/`test_fstring`/`test_tokenize`. Trying the whole test suite (is there a specific option I might be missing?)

**Comment by lysnikolaou** [comment]
More info. When running it with Python, I get the following error:

```python3
root@9ee555036b0f:/usr/src/cpython# cat t.py
eval('f"a"')
root@9ee555036b0f:/usr/src/cpython# ./python t.py
Fatal Python error: tok_backup: tok_backup: wrong character
Python runtime state: initialized

Current thread 0x0000ffff9de38750 (most recent call first):
  File "/usr/src/cpython/t.py", line 1 in <module>
Aborted
```

Here's a simple step though `tok_get_fstring_mode` on gdb in the last pass that generates the error
```
(gdb) file ./python
Reading symbols from ./python...
warning: File "/usr/src/cpython/python-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load".
To enable execution of this file add
	add-auto-load-safe-path /usr/src/cpython/python-gdb.py
line to your configuration file "/root/.gdbinit".
To completely disable this security protection add
	set auto-load safe-path /
line to your configuration file "/root/.gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
	info "(gdb)Auto-loading safe path"
(gdb) break tok_get_fstring_mode
Breakpoint 1 at 0x17119c: file Parser/tokenizer.c, line 2442.
(gdb) r t.py
Starting program: /usr/src/cpython/python t.py
warning: Error disabling address space randomization: Operation not permitted
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".

Breakpoint 1, tok_get_fstring_mode (tok=0xaaab0ecfec60, current_tok=0xaaab0ecff7d0, token=0xffffc95f8278) at Parser/tokenizer.c:2442
2442	{
(gdb) c
Continuing.

Breakpoint 1, tok_get_fstring_mode (tok=0xaaab0ecfec60, current_tok=0xaaab0ecff7d0, token=0xffffc95f8278) at Parser/tokenizer.c:2442
2442	{
(gdb) p tok->cur
$1 = 0xffff883fc1e3 "\""
(gdb) p tok->buf
$2 = 0xffff883fc1e0 "f\"a\""
(gdb) n
2448	    tok->start = tok->cur;
(gdb)
2449	    tok->first_lineno = tok->lineno;
(gdb)
2450	    tok->starting_col_offset = tok->col_offset;
(gdb)
2454	    char start_char = tok_nextc(tok);
(gdb)
2455	    char peek1 = tok_nextc(tok);
(gdb) p start_char
$3 = 34 '"'
(gdb) s
tok_nextc (tok=0xaaab0ecfec60) at Parser/tokenizer.c:1169
1169	{
(gdb) n
1172	        if (tok->cur != tok->inp) {
(gdb)
1176	        if (tok->done != E_OK) {
(gdb)
1179	        if (tok->fp == NULL) {
(gdb)
1180	            rc = tok_underflow_string(tok);
(gdb) s
tok_underflow_string (tok=0xaaab0ecfec60) at Parser/tokenizer.c:965
965	tok_underflow_string(struct tok_state *tok) {
(gdb) list
960	    } while (tok->inp[-1] != '\n');
961	    return 1;
962	}
963
964	static int
965	tok_underflow_string(struct tok_state *tok) {
966	    char *end = strchr(tok->inp, '\n');
967	    if (end != NULL) {
968	        end++;
969	    }
(gdb)
970	    else {
971	        end = strchr(tok->inp, '\0');
972	        if (end == tok->inp) {
973	            tok->done = E_EOF;
974	            return 0;
975	        }
976	    }
977	    if (tok->start == NULL) {
978	        tok->buf = tok->cur;
979	    }
(gdb) n
966	    char *end = strchr(tok->inp, '\n');
(gdb)
967	    if (end != NULL) {
(gdb)
971	        end = strchr(tok->inp, '\0');
(gdb)
972	        if (end == tok->inp) {
(gdb)
973	            tok->done = E_EOF;
(gdb)
974	            return 0;
(gdb)
tok_nextc (tok=0xaaab0ecfec60) at Parser/tokenizer.c:1189
1189	        if (tok->debug) {
(gdb) list
1184	        }
1185	        else {
1186	            rc = tok_underflow_file(tok);
1187	        }
1188	#if defined(Py_DEBUG)
1189	        if (tok->debug) {
1190	            fprintf(stderr, "line[%d] = ", tok->lineno);
1191	            print_escape(stderr, tok->cur, tok->inp - tok->cur);
1192	            fprintf(stderr, "  tok->done = %d\n", tok->done);
1193	        }
(gdb)
1194	#endif
1195	        if (!rc) {
1196	            tok->cur = tok->inp;
1197	            return EOF;
1198	        }
1199	        tok->line_start = tok->cur;
1200
1201	        if (contains_null_bytes(tok->line_start, tok->inp - tok->line_start)) {
1202	            syntaxerror(tok, "source code cannot contain null bytes");
1203	            tok->cur = tok->inp;
(gdb) n
1195	        if (!rc) {
(gdb)
1196	            tok->cur = tok->inp;
(gdb)
1197	            return EOF;
(gdb)
tok_get_fstring_mode (tok=0xaaab0ecfec60, current_tok=0xaaab0ecff7d0, token=0xffffc95f8278) at Parser/tokenizer.c:2456
2456	    tok_backup(tok, peek1);
(gdb) p peek1
$4 = 255 '\377'
```

**Comment by isidentical** [comment]
> When the f-string is too small, it results in either start_char or peek1 or both [here](https://github.com/pablogsal/cpython/blob/fstring-grammar-rebased-after-sprint/Parser/tokenizer.c#L2454..L2455) to be EOF.

Oh, this kind of makes sense. At least on how we got there. I wonder whether we could simply look at the `peek1` if the `start_char` is `{`/`}`. This would prevent the secondary `tok_nextc`/`tok_backup` pair when in case the string is too small.

**Comment by sunmy2019** [comment]
> The big questions to me is how do we end up with `peek1 == 255`, when it very clearly came from `return EOF` and the subsequent check `c == EOF` in `tok_backup` fails.

`c` is of type char, `-1` is of type int. What will the compiler choose when comparing with `c == (-1)`?

**The answer should be "they are lifted to int."**

The tricky part is that char is unsigned on those platforms (arm). Thus, `c == -1` never evaluates to true.

I am able to reproduce this problem on x86 with

```
   unsigned char start_char = tok_nextc(tok);
   unsigned char peek1 = tok_nextc(tok);
```

**Comment by sunmy2019** [comment]
C allows any `int` to convert to `char`, which may silently change its value. It is exactly what we have in this case.

explicitly using `signed char` or `int` should fix this 255 problem. (I prefer the `int` one)

**Comment by lysnikolaou** [comment]
Oooh, that's right! Didn't know that ARM has unsigned chars by default. Pushed a fix.

**Comment by isidentical** [comment]
Wow, thats a nice find!!

**Comment by lysnikolaou** [comment]
Ι can confirm that `test_fstring` passes without problems on my machine that previously had issues.

**Comment by pablogsal** [comment]
@isidentical @lysnikolaou I plan to land this bad boy as this is already gigantic. Please, take a last look and LGTM as hard as you can :)

**Comment by lysnikolaou** [comment]
@pablogsal I'll take a last look a bit later today. Would it make sense to run a test on PyPI packages (top 100 is probably enough) before doing so?

**Comment by pablogsal** [comment]
> Would it make sense to run a test on PyPI packages (top 100 is probably enough) before doing so?

Excellent idea! IIRC @isidentical has infrastructure already to check this

**Comment by sunmy2019** [comment]
I can also run some checks. Is the source code of @isidentical's infrastructure available?

**Comment by lysnikolaou** [comment]
> I can also run some checks. Is the source code of @isidentical's infrastructure available?

I think that @isidentical is using [isidentical/sourced](https://github.com/isidentical/sourced) to do that. There's also a [couple of scripts for this on the pegen project](https://github.com/we-like-parsers/pegen/tree/main/scripts) (we were using those while developing the parser), but Batuhan's project is much better for this.

**Comment by pablogsal** [comment]
> > I can also run some checks. Is the source code of @isidentical's infrastructure available?
> 
> I think that @isidentical is using [isidentical/sourced](https://github.com/isidentical/sourced) to do that. There's also a [couple of scripts for this on the pegen project](https://github.com/we-like-parsers/pegen/tree/main/scripts) (we were using those while developing the parser), but Batuhan's project is much better for this.

As a quick experiment I have used https://github.com/we-like-parsers/pegen/blob/main/scripts/download_pypi_packages.py to run a quick analysis using this (really really really dirty) script:

```
import ast
import os
import subprocess

directory_path = "."

def compare_python_interpreters(file_path):
    with open(file_path, "r") as f:
        code = f.read()

    try:
        # Parse the code using the first Python interpreter
        ast1 = ast.dump(ast.parse(code))
    except:
        ast1 = None

    try:
        # Parse the code using the second Python interpreter
        cmd = ["/Users/pgalindo3/github/python/main/python.exe", "-c", f"import ast; code = (open(\"{file_path}\", "r").read()); print(ast.dump(ast.parse(code)))"]
        output = subprocess.check_output(cmd, stderr=subprocess.DEVNULL, shell=False)
        ast2 = output.decode().strip()
    except subprocess.CalledProcessError as e:
        ast2 = None

    return ast1 == ast2

for root, dirs, files in os.walk(directory_path):
    for file in files:
        if file.endswith(".py"):
            file_path = os.path.join(root, file)
            result = compare_python_interpreters(file_path)
            print(f"Checking {file_path}")
            if not result:
                raise ValueError(file_path)
```

**Comment by sunmy2019** [comment]
I sampled 99 projects on pypi-popular using https://github.com/isidentical/sourced
tested on 9602 files.
parse->unparse: all passed.
parse->dump and compare with py3.10.6: all passed

**Comment by lysnikolaou** [comment]
That's great! Thanks a lot, @sunmy2019!

**Comment by pablogsal** [comment]
Hummm, looks we stil have some buildbots failing:

```
Parser/tokenizer.c:2465:13: runtime error: index 2 out of bounds for type 'int[2]'
```

**Comment by sunmy2019** [comment]
I prefer the no `u` version, which describes the kind of each component more precisely and loses less info.

But I am okay with both since I don't find a real-world case where this field really matters.

**Comment by lysnikolaou** [comment]
No strong opinion here. I'm okay going either way.

**Comment by pablogsal** [comment]
We have a missing bug to solve, catched by the UB buildbot:

```

test_many_expressions (test.test_fstring.TestCase.test_many_expressions) ... ok
test_misformed_unicode_character_name (test.test_fstring.TestCase.test_misformed_unicode_character_name) ... ok
Parser/tokenizer.c:2321:20: runtime error: index -1 out of bounds for type 'int[3]'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior Parser/tokenizer.c:2321:20 in 
make: *** [Makefile:1954: buildbottest] Error 1
```

**Comment by lysnikolaou** [comment]
Pushed a fix for this. It seems that the check was simple, so a few more sets of eyes wouldn't hurt!

**Comment by pablogsal** [comment]
> Pushed a fix for this. It seems it was simple, so a few more sets of eyes wouldn't hurt!

Maybe we can add a bunch of assets here and there so we don't need the usan buildbots to detect this cases.

I will try to add some macros maybe.

**Comment by sunmy2019** [comment]
@pablogsal @lysnikolaou This issue isn't fixed.

With 367d7f6 and ubsan on, the following fails:

```py
class A:
    def __format__(self, __format_spec: str) -> str:
        return ""
f"{A(): {A():{A(): {1}}}}"
```

```
Parser/tokenizer.c:2535:42: runtime error: index 3 out of bounds for type 'int [3]'
Parser/tokenizer.c:2321:45: runtime error: index 3 out of bounds for type 'int [3]'
Parser/tokenizer.c:2423:84: runtime error: index 3 out of bounds for type 'int [3]'
```

**Comment by sunmy2019** [comment]
Technically, we should check every time we change this index.

By the way, do you have any thoughts on https://github.com/python/cpython/pull/102855#issuecomment-1506757771

> One issue is that, with current grammar
> 
> ```python
> f"{lambda x:{123}}"
> ```
> 
> will be recognized as a valid lambda

Will this be fixed?

**Comment by pablogsal** [comment]
> With [367d7f6](https://github.com/python/cpython/commit/367d7f6f5bda53cc9f586a8f713de7bb1ea91ccf) and ubsan on, the following fails:

I cannot reproduce this on my macOS machine. I am pushing also a commit to add a bunch of asserts.

**Comment by pablogsal** [comment]
> Will this be fixed?

We need to discuss this later, but surely this won't be fixed in this PR

**Comment by pablogsal** [comment]
> > With [367d7f6](https://github.com/python/cpython/commit/367d7f6f5bda53cc9f586a8f713de7bb1ea91ccf) and ubsan on, the following fails:
> 
> I cannot reproduce this on my macOS machine. I am pushing also a commit to add a bunch of asserts.

The asserts I added fail even when compiling the interpreter so there is certainly something going on here.

File: Parser/tokenizer.c [hunk 4]
```diff
@@ -2462,9 +2485,9 @@ tok_get_fstring_mode(struct tok_state *tok, tokenizer_mode* current_tok, struct
                 return MAKE_TOKEN(syntaxerror(tok, "f-string: expressions nested too deeply"));
             }
 
-            current_tok->bracket_mark[++current_tok->bracket_mark_index] = current_tok->bracket_stack;
+            *TOK_GET_BRACKET_MARK(current_tok) = current_tok->bracket_stack;
```
Comment (sunmy2019) [inline Parser/tokenizer.c:2486]: `++` is missing

File: Parser/tokenizer.c [hunk 5]
```diff
@@ -2529,8 +2552,9 @@ tok_get_fstring_mode(struct tok_state *tok, tokenizer_mode* current_tok, struct
             if (peek != '{' || in_format_spec) {
                 tok_backup(tok, peek);
                 tok_backup(tok, c);
-                current_tok->bracket_mark[++current_tok->bracket_mark_index] = current_tok->bracket_stack;
-                tok->tok_mode_stack[tok->tok_mode_stack_index].kind = TOK_REGULAR_MODE;
+                current_tok->bracket_mark_index++;
+                *TOK_GET_BRACKET_MARK(current_tok) = current_tok->bracket_stack;
+                TOK_GET_MODE(tok)->kind = TOK_REGULAR_MODE;
```
Comment (sunmy2019) [inline Parser/tokenizer.c:2558]: `->` has higher precedence over `&`. Paren the macro definition in no debug mode.

File: Parser/tokenizer.c [hunk 6]
```diff
@@ -2003,6 +2176,67 @@ tok_get(struct tok_state *tok, struct token *token)
...
+                break;
+            case 'R':
+            case 'r':
+                current_tok->f_string_raw = 1;
+                break;
+            default:
+                Py_UNREACHABLE();
+        }
+
+        current_tok->bracket_stack = 0;
```
Comment (sunmy2019) [inline Parser/tokenizer.c:2234]: This should still be `-1`. Otherwise, a lot of things need to be changed consequently.

**Review by sunmy2019 — CHANGES_REQUESTED** [review]

File: Parser/tokenizer.c (hunk 6 above)
Comment (pablogsal) [reply Parser/tokenizer.c:2234]: Hummm, I am not going to change this as I just want to add the macros.

Comment (sunmy2019) [reply Parser/tokenizer.c:2234]: Sorry, I have commented on the wrong line. Already fixed now.

**Comment by pablogsal** [comment]
@isidentical @lysnikolaou please check the pr again and LGTM if you are ready to land

**Comment by pablogsal** [comment]
> One more nit, why is this `MAX_EXPR_NEXTING` instead of `MAX_EXPR_NESTING`?

Ah that's a typo 👍

**Comment by isidentical** [comment]
Gave an initial pass, everything seems to be looking great 🚀

File: Parser/tokenizer.c [hunk 7]
```diff
@@ -43,6 +43,28 @@
...
+    return &(tok->tok_mode_stack[tok->tok_mode_stack_index]);
+}
+static inline tokenizer_mode* TOK_NEXT_MODE(struct tok_state* tok) {
+    assert(tok->tok_mode_stack_index >= 0);
+    assert(tok->tok_mode_stack_index < MAXLEVEL);
+    return &(tok->tok_mode_stack[++tok->tok_mode_stack_index]);
+}
+static inline int *TOK_GET_BRACKET_MARK(tokenizer_mode* mode) {
+    assert(mode->bracket_mark_index >= 0);
+    assert(mode->bracket_mark_index < MAX_EXPR_NEXTING);
```
Comment (pablogsal) [inline Parser/tokenizer.c]:
```suggestion
    assert(mode->bracket_mark_index < MAX_EXPR_NESTING);
```

File: Parser/tokenizer.c [hunk 8]
```diff
@@ -2162,6 +2458,187 @@ tok_get(struct tok_state *tok, struct token *token)
...
+    // before it.
+    int start_char = tok_nextc(tok);
+    int peek1 = tok_nextc(tok);
+    tok_backup(tok, peek1);
+    tok_backup(tok, start_char);
+
+    if ((start_char == '{' && peek1 != '{') || (start_char == '}' && peek1 != '}')) {
+        if (start_char == '{') {
+            current_tok->bracket_mark_index++;
+            if (current_tok->bracket_mark_index >= MAX_EXPR_NEXTING) {
```
Comment (pablogsal) [inline Parser/tokenizer.c]:
```suggestion
            if (current_tok->bracket_mark_index >= MAX_EXPR_NESTING) {
```

File: Parser/tokenizer.h [hunk 9]
```diff
@@ -33,6 +33,31 @@ struct token {
     const char *start, *end;
 };
 
+enum tokenizer_mode_kind_t {
+    TOK_REGULAR_MODE,
+    TOK_FSTRING_MODE,
+};
+
+#define MAX_EXPR_NEXTING 3
```
Comment (pablogsal) [inline Parser/tokenizer.h]:
```suggestion
#define MAX_EXPR_NESTING 3
```

File: Parser/tokenizer.h [hunk 10]
```diff
@@ -33,6 +33,31 @@ struct token {
...
+
+typedef struct _tokenizer_mode {
+    enum tokenizer_mode_kind_t kind;
+
+    int bracket_stack;
+    int bracket_mark[MAX_EXPR_NEXTING];
```
Comment (pablogsal) [inline Parser/tokenizer.h]:
```suggestion
    int bracket_mark[MAX_EXPR_NESTING];
```

File: Parser/tokenizer.c [hunk 11]
```diff
@@ -2162,6 +2458,187 @@ tok_get(struct tok_state *tok, struct token *token)
...
+        }
+
+        int in_format_spec = current_tok->last_expr_end != -1 && current_tok->bracket_mark_index >= 0;
+        if (c == '{') {
+            int peek = tok_nextc(tok);
+            if (peek != '{' || in_format_spec) {
+                tok_backup(tok, peek);
+                tok_backup(tok, c);
+                current_tok->bracket_mark_index++;
+                if (current_tok->bracket_mark_index >= MAX_EXPR_NEXTING) {
```
Comment (pablogsal) [inline Parser/tokenizer.c]:
```suggestion
                if (current_tok->bracket_mark_index >= MAX_EXPR_NESTING) {
```

**Comment by pablogsal** [comment]
> Gave an initial pass, everything seems to be looking great 🚀  

Do you mind approving it formally?

**Comment by pablogsal** [comment]
Same thing for @lysnikolaou 😉

**Review by lysnikolaou — APPROVED** [review]
LGTM! Let's merge! 🚀

**Review by isidentical — APPROVED** [review]
💫 Looks great, thanks everyone for their amazing work!!

**Review by rllane17 — COMMENTED** [review]
need help with getting the program to run

**Pull request merged**